Project Cluedo - Jamie Cook

the-j-curl opened the pull request:
I completed the blue requirements. I also completed all of the red level goals. Which included:
Adding favouriteWeapon property, and making this a random selection. 
I also added a loading animation.

I had trouble trying to get the shuffleFavouriteWeapon function working. And I also spent a lot of time getting the animation to work. I found that after I added the animation my RESET button stopped working. I have left this in my product to hopefully get some feedback on. 
Another strange issue I came across was when declaring my revealMystery function it only worked when I created it using the function keyword, and did not execute if I declared it as: const revealMystery = () => { 
(see line 334 in script.js). I posted this as a question on SO FYI.

I also feel a little confused with the JavaScript once the code starts looking more complex. 
I would of liked to have had more time so I could have created a new array as part of the black level goals.

the-j-curl committed:
objects created for the suspects and the weapons

the-j-curl committed:
Step 3 complete. Weapon images added to assests and JS file

the-j-curl committed:
step 4 complete. Styled cards and background

the-j-curl committed:
background movement fixed. upload to netlify

the-j-curl committed:
animation added to card, but not complete

the-j-curl committed:
favourite weapon function added, animation added

the-j-curl committed:
readme notes updated

the-j-curl committed:
comment edited on line 334

the-j-curl committed:
changed function declaration for clearMystery. this is now working

the-j-curl committed:
changed picked variables to false so game can be replayed after being reset

ylvan75 commented on code/clue.js:
```diff
@@ -1,59 +1,363 @@
-// THIS FUNCTION WILL RANDOMLY SELECT ONE ITEM FROM THE ARRAY THAT YOU PASS IN TO THE FUNCTION.
-// YOU DON'T NEED TO CHANGE THIS, JUST TRY TO UNDERSTAND IT. AND HOW TO USE IT.
 const randomSelector = array => {
   return array[Math.floor(Math.random() * array.length)]
 }
```
Missing semi-colon

ylvan75 commented on code/clue.js:
```diff
@@ -1,59 +1,363 @@
-    'killerName'
-  ).innerHTML = `${mystery.killer.firstName} ${mystery.killer.lastName}`
+    if (!killerIsPicked) {
+    document.getElementById('killerLoader').style.visibility='visible' 
```
Missing semi-colon

ylvan75 commented on code/clue.js:
```diff
@@ -1,59 +1,363 @@
+      mystery.room = randomSelector(rooms);
+      document.getElementById('roomName').innerHTML = mystery.room.name;
+      document.getElementById('roomImage').src = mystery.room.image;
+      roomIsPicked = true
```
Missing semi-colon

ylvan75 commented on code/clue.js:
```diff
@@ -1,59 +1,363 @@
+    }
+};
+
+document.getElementById('killerCard').addEventListener('click', pickKiller);
```
You don't need this line since you are using "onclick" in the HTML-file. Either you keep this in the JS-file and remove the "onclick" or vice versa.

Please update you PR and I will later on approve it!

ylvan75 commented on code/clue.js:
```diff
@@ -1,59 +1,363 @@
 
-// CREATE FUNCTIONS pickWeapon and pickRoom in a similar way.
+};
+document.getElementById('weaponCard').addEventListener('click', pickWeapon);
```
You don't need this line since you are using "onclick" in the HTML-file. Either you keep this in the JS-file and remove the "onclick" or vice versa.

Please update you PR and I will later on approve it!

the-j-curl replied:
Thanks - well spotted. I have removed the eventListener

ylvan75 commented on code/clue.js:
```diff
@@ -1,59 +1,363 @@
+  }
+};
+
+document.getElementById('roomcard').addEventListener('click', pickRoom);
```
You don't need this line since you are using "onclick" in the HTML-file. Either you keep this in the JS-file and remove the "onclick" or vice versa.

If you want to keep it in the JS-file you need to change `roomcard` to `roomCard` to get it to work. If you goto your site and look at the Console you will see that it complaining about this row.

Please update you PR and I will later on approve it!

ylvan75 commented on code/clue.js:
```diff
@@ -1,59 +1,363 @@
-  ).innerHTML = `${mystery.killer.firstName} ${mystery.killer.lastName}`
+    if (!killerIsPicked) {
+    document.getElementById('killerLoader').style.visibility='visible' 
+    setTimeout(function () {
```
Use arrow notation for functions `setTimeout(() => {`

ylvan75 commented on code/clue.js:
```diff
@@ -1,59 +1,363 @@
+const pickWeapon = () => {
+  if (!weaponIsPicked) {
+    document.getElementById('weaponLoader').style.visibility='visible';
+    setTimeout(function () {
```
Use arrow notation for functions `setTimeout(() => {`

ylvan75 commented on code/clue.js:
```diff
@@ -1,59 +1,363 @@
+const pickRoom = () => {
+    if (!roomIsPicked) {
+    document.getElementById('roomLoader').style.visibility='visible';
+    setTimeout(function () {
```
Use arrow notation for functions `setTimeout(() => {`

ylvan75 commented on code/clue.js:
```diff
@@ -1,59 +1,363 @@
+
+//const revealMystery = () => { - this function does not work when created using const keyword
+  function revealMystery() {
+    if (killerIsPicked === true && weaponIsPicked === true && roomIsPicked ===true) {
+      document.getElementById('mystery').innerHTML = `The murder was committed by ${mystery.killer.firstName} ${mystery.killer.lastName}, in the ${mystery.room.name} with the ${mystery.weapon.name}`;
```
A tip if you would have wanted a message to be presented for the user if you try to click "Reveal the crime" before all deck cards where picked.

```suggestion
  if (
    mystery.killer === '' ||
    mystery.weapon === '' ||
    mystery.room === ''
  ) {
    document.getElementById("mystery").innerHTML =
      "Click all cards above to reveal the crime!";
  } else {
    document.getElementById(
      "mystery"
    ).innerHTML = `The murder was commited by ${mystery.killer.firstName} ${mystery.killer.lastName}, in the ${mystery.room} room with a ${mystery.weapon.name}!`;
```

ylvan75 commented on code/clue.js:
```diff
@@ -1,59 +1,363 @@
+// Reveal function
+
+//const revealMystery = () => { - this function does not work when created using const keyword
+  function revealMystery() {
```
As you have mentioned you want to use arrow notation for functions but it didn't work for you. I don't see what you need to do to get it to work. Wondering if it can have something to do that you do onclick on a html-tag ` <button>`

ylvan75 commented on code/clue.js:
```diff
@@ -1,59 +1,363 @@
 
-// STEP 4 - CREATE A FUNCTION revealMystery that will be invoked when you click that button. It should show something like:
-// 'The murder was committed by Jacob Green, in the living room with a rope.'
+function clearMystery() {
```
As you have mentioned you want to use arrow notation for functions but it didn't work for you. I don't see what you need to do to get it to work. Wondering if it can have something to do that you do onclick on a html-tag ` <button>`

ylvan75 commented on code/index.html:
```diff
@@ -4,34 +4,50 @@
-        <img id="killerImage" src="" height="100" />
         <p id="killerDescription"></p>
+        <p id="favouriteWeapon"></p>
+        <div id="killerLoader" class="loader"></div>
```
I'm just curious, when is this one used?

ylvan75 commented on code/index.html:
```diff
@@ -4,34 +4,50 @@
+        </div>
         <p id="weaponName"></p>
         <p id="weaponWeight"></p>
+        <div id="weaponLoader" class="loader"></div>
```
I'm just curious, when is this one used?

ylvan75 commented on code/index.html:
```diff
@@ -4,34 +4,50 @@
+          <img id="roomImage" class="room-image" src="" />
+        </div>
         <p id="roomName"></p>
+        <div id="roomLoader" class="loader"></div>
```
I'm just curious, when is this one used?

the-j-curl replied:
This div is my card loader animation

ylvan75 commented on code/style.css:
```diff
@@ -26,6 +67,18 @@ p {
+  animation: 0.8s 1 change-colour;
+}
+
+.killer-deck, .weapon-deck, .room-deck, .heading {
```
Good that you group them like this, since they use same styling

ylvan75 commented on code/style.css:
```diff
@@ -26,6 +67,18 @@ p {
+  font-weight: bold;
+}
+
+.killer-deck, .weapon-deck, .room-deck, p {
```
Good that you group them like this, since they use same styling

ylvan75 commented on code/style.css:
```diff
@@ -35,24 +88,133 @@ p {
+  font-weight: bold;
 }
+
+.button-reveal:hover {
+  opacity: 0.9;
+  cursor: pointer;
```
Nice effect!

ylvan75 commented on code/style.css:
```diff
@@ -35,24 +88,133 @@ p {
+  cursor: pointer;
+}
+
+.button-reset:hover {
+  opacity: 0.9;
+  cursor: pointer;
```
Nice effect!

ylvan75 reviewed (changes requested):
Great job Jamie :-)

I really liked your Pull Request description (got me inspired to make mine better in the future). It gave me a good overview of what level you aimed for and what you have been struggling with during the project. Nice that you have highlighted things as the RESET button and revealMystery function. Made it easier for me when I started to test your website and later on looking into the code.

It was also good that you in the README said that you focused om the JavaScript and didn't put in time for the responsiveness part. Good decision!

I like your background image and that you have added images for weapon and room, the deck cards looks more balanced then.

I notice when I tested your site that I didn't get any message with instructions what to do if I tried to "Reveal the crime" before all deck cards where picked. In your JS-file at the function revealMystery I gave you a tip on how you could have done it, if that would have been something you wanted to do.

Don't forget to be consistent if you want to use semi-colon. I found a couple of places in the JS-file that the semi-colon is missing, some examples line 257, 258, 279 and 324.

All three HTML, CSS and JS-files are really clean and made it easy to review.

Before being approved decide if you wanna use `onclick` in the HTML-file or `addEventListener` in JS-file, see my comment in the JS-file on line 296, 313 and 329.

the-j-curl committed:
missing semi colon added, eventListener removed which wasn't being used

the-j-curl commented:
Thanks for the feedback Ylva. Some useful comments.
I have added the missing semi-colons to be consistent and removed the event Listener which wasn't being used.

ylvan75 commented:
Thumbs up from me!

EmOlsz commented on code/clue.js:
```diff
@@ -1,59 +1,357 @@
-// STEP 1 - CREATE OBJECTS FOR ALL THE SUSPECTS, SOMETHING LIKE THIS:
+let killerIsPicked = false;
+let weaponIsPicked = false;
+let roomIsPicked = false;
```
Good idea to store this info in variables! Nonetheless, one thing could be improved :
There is a strict convention about boolean variables names. They should always start with `is` as a prefix: `isKillerPicker`, `isWeaponPicked`, `isRoomPicked`

EmOlsz commented on code/clue.js:
```diff
@@ -1,59 +1,357 @@
+  favouriteWeapon: 'rope'
+};
+
+let professorPlum = {
```
You don't overwrite those values anywhere in app, they should be declared with `const`

EmOlsz commented on code/clue.js:
```diff
@@ -1,59 +1,357 @@
+// Mystery object
+let mystery = {
+  killer: '',
+  weapon: '',
```
It's even more semantic approach to initialise mystery object with `null` values like this :

```
const mystery = {
  killer: null,
  room: null,
  weapon: null
}
```

Moreover, you can use `const` keyword - we will be changing values of properties, that's true, but JavaScript will not consider this as overwriting. What counts for JavaScript is that object will stay object

EmOlsz commented on code/clue.js:
```diff
@@ -1,59 +1,357 @@
+const shuffleFavouriteWeapon = () => {
+  suspects.forEach((suspect) => {
+  const randomWeapon = randomSelector(weapons);
+  suspect.favouriteWeapon = randomWeapon.id;
```
wrong indentation

EmOlsz commented on code/clue.js:
```diff
@@ -1,59 +1,357 @@
-  document.getElementById(
-    'killerName'
-  ).innerHTML = `${mystery.killer.firstName} ${mystery.killer.lastName}`
+    if (!killerIsPicked) {
```
very nice Jamie :)

EmOlsz commented on code/clue.js:
```diff
@@ -1,59 +1,357 @@
+
+//const revealMystery = () => { - this function does not work when created using const keyword
+  function revealMystery() {
+    if (killerIsPicked === true && weaponIsPicked === true && roomIsPicked ===true) {
```
Or we can simplify that statement :

```
if (killerIsPicked && weaponIsPicked && roomIsPicked) { ... }
```

EmOlsz commented on code/style.css:
```diff
@@ -1,19 +1,60 @@
 body {
   margin: 0;
   font-family: "Courier New", Courier, monospace;
+  background-image: url(./assets/image-main-dark5.jpg);
```
It's far more safe approach to wrap path in parentheses. Without them, some browsers might not find the path
`url("./assets/image-main-dark5.jpg");`

EmOlsz commented on code/style.css:
```diff
@@ -35,24 +88,133 @@ p {
 
-button {
-  padding: 8px 16px;
+.button-reveal {
```
I'm happy you use classes only in your CSS. In general, your CSS code is very clean, great job :)

EmOlsz commented on code/style.css:
```diff
@@ -35,24 +88,133 @@ p {
+@keyframes bounce {
+  0% {
+    width: 16px;
+    height: 16px;
```
These properties never change, no need to add them in `@keyframes`. The best approach would be to add them to `.loader` class

EmOlsz reviewed (approved):
I'm very happy about quality of your code Jamie! It's well thought, consequent and aesthetic (however indentation could be improved). Especially CSS part caught my attention :)
Take a look on my comments to improve even more, nonetheless project is approved!

the-j-curl commented:
@EmOlsz - Thanks for all the feedback Maks. I have updated my code as per your advice and will carry this info over to future projects.

the-j-curl committed:
made site responsive cleaned up code as per code review

the-j-curl committed:
Added else statement to revealMystery function

the-j-curl committed:
replaced empty array value  with null value

the-j-curl committed:
updated readme
